src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
     errors.into_result(())
 }
 
+// Go through all functions and check if they are castable
+// based only on their args
+pub fn initialize_castable(functions: &mut FunctionMap, types: &TypeMap) {
+    for func in functions.values_mut() {
+        for arg in &func.args {
+            if arg.param_type.is_associated_resource(types) {
+                func.is_castable = false;
+
+                // If we have found one associated resource
+                // we can continue
+                continue;
+            }
+        }
+    }
+}
+
+// Go through all of the functions and check if they are castable
+// base on functions they call.
+pub fn determine_castable(functions: &mut FunctionMap, types: &TypeMap) -> u64 {
+    let mut num_changed: u64 = 0;
+    // We need tmp_functions to avoid a immutable borrow after a mutable one.
+    let tmp_functions = functions.clone();
```

Comment:
This is probably a follow-up task since we want to get this working functionally, but this sounds really non-performant.  Since determine_castable() is called in a loop and functions is reasonably sizable, this is a ton of copying.

A future optimization could save a list of names to mark uncastable and then mark them as a cleanup task so the whole below step is done immutably.

I recently filed an issue in our internal issue tracker to add performance benchmarking.  That would be helpful if we go back to optimize so that we can see the impact of a proposed performance optimization like that.

Reply:
https://github.com/dburgener/cascade/issues/138